[CK-Tile] Fix quant example code

amd-khushbu opened the pull request:
## Proposed changes

With the recent changes on develop branch we had some duplicate files for preshuffle AQuant and BQuant.
With this PR, Generalizing the example to have Aquant/BQuant/RowColQuant quantization modes.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [X] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

amd-khushbu committed:
initial commit

amd-khushbu committed:
remove extra files

amd-khushbu committed:
fixing errors

amd-khushbu committed:
updated ReadMe file for mapping of diff quants with diff configs

amd-khushbu committed:
Merge branch 'develop' into fix_quant_example

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_quant_basic.cpp:
```diff
@@ -18,58 +18,49 @@ template <typename GemmConfig,
+        ck_tile::BaseGemmPipelineAgBgCrCompV3<GemmPipelineProblem>; // TODO :: check for BQuant
 
-    const ck_tile::index_t K_split      = (args.K + K_Tile - 1) / K_Tile * K_Tile;
+    // using BaseGemmPipeline = typename PipelineTypeTraits<
```
Extra Comments?

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_quant_basic.cpp:
```diff
@@ -80,36 +71,52 @@ float gemm_calc_quant(const ck_tile::QuantGemmHostArgs& args, const ck_tile::str
-                               ck_tile::GemmPipelineAgBgCrCompV3<CodegenPipelineProblem>>;
+                               ck_tile::AQuantGemmPipelineAgBgCrCompV3<PipelineProblem>,
+                               ck_tile::BQuantGemmPipelineAgBgCrCompV3<PipelineProblem>>>;
+        // using GemmPipeline = typename PipelineTypeTraits<GemmConfig::Pipeline>::template
```
Extra Comment

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_utils.hpp:
```diff
@@ -206,25 +267,25 @@ auto create_args(int argc, char* argv[])
         .insert("a_layout", "R", "A tensor data layout - Row by default")
         .insert("aq_layout", "R", "Aq tensor data layout - Row by default")
         .insert("b_layout", "C", "B tensor data layout - Column by default")
+        .insert("bq_layout", "C", "Bq tensor data layout - Column by default")
```
Do we need to have a aq or bq layout? We could make the quant layout in the same as the a or b

amd-khushbu replied:
Do you want me to remove both aq_layout and bq_layout ?

ThomasNing replied:
Yes please

ThomasNing replied:
We could do that in the next PR.

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_utils.hpp:
```diff
@@ -11,9 +11,8 @@
-#define CK_TILE_PIPELINE_PREFILL 1
-#define CK_TILE_PIPELINE_DECODE 2
-#define CK_TILE_PIPELINE_PRESHUFFLEQUANT 3
+#define CK_TILE_AQUANT_PIPELINE 1
```
I think we could delete this part.

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_utils.hpp:
```diff
@@ -122,11 +164,30 @@ struct GemmConfigPreshuffleQuant : public GemmConfigBase
     static constexpr ck_tile::index_t K_Warp_Tile =
         get_k_from_preshuffled_warp_tile<PrecType, M_Warp_Tile>();
 
-    static constexpr int kBlockPerCu = 1;
```
Let's put the kBlockPerCu as 1 in the base as default for now.

amd-khushbu replied:
i have moved kBlockPerCu to base struct and initialized it to 1.

ThomasNing commented on example/ck_tile/38_block_scale_gemm/gemm_utils.hpp:
```diff
@@ -89,7 +98,7 @@ struct GemmConfigBase
 
 template <typename PrecType>
-struct GemmConfigQuant : public GemmConfigBase
+struct GemmConfigAQuant : public GemmConfigBase
```
If we are sharing over 1 config then we can merge the AQuant and BQuant. Grouped block scale quant with preshuffle and transpose C can be the separate config in that case.

ThomasNing commented on example/ck_tile/38_block_scale_gemm/run_gemm_quant_example.inc:
```diff
@@ -5,28 +5,21 @@
-
 template <typename T>
-auto shuffle_aq(const ck_tile::HostTensor<T>& t, int block_aq_k)
+auto shuffle_aq(const ck_tile::HostTensor<T>* t, int block_aq_k)
```
Could I know why we change to pass in a pointer here?

amd-khushbu replied:
From the calling function, HostTensor is now a unique pointer because it can be null_ptr for bquant.

AviralGoelAMD commented on example/ck_tile/38_block_scale_gemm/gemm_quant_basic.cpp:
```diff
@@ -274,6 +290,14 @@ int run_gemm_example(int argc, char* argv[])
         }
+        else if(quant_mode == "bquant")
+        {
+            return run_gemm_example_prec_type<GemmConfig<ck_tile::fp8_t>,
```
```suggestion
            return run_gemm_example_prec_type<GemmConfig<ck_tile::bf8_t>,
```
perhaps this?

amd-khushbu committed:
addressing review comments

AviralGoelAMD commented:
1. CMakeLists.txt file needs to be updated.
2. Remove the `void print_help(const char* program_name)`. We already put this information in README.md
3. https://github.com/ROCm/composable_kernel/pull/2813/files#diff-3ad7f4d6fb884627dcc872a14043585e024bc46967307c858d67c7abe5f5943cL262 and https://github.com/ROCm/composable_kernel/pull/2813/files#diff-3ad7f4d6fb884627dcc872a14043585e024bc46967307c858d67c7abe5f5943cL287
bquant is also supported here, we can add it to error message string
```cpp 
else   
{       
throw std::runtime_error("Unsupported quantization mode! Use 'aquant' or 'rowcol'");   
}
```
5. https://github.com/ROCm/composable_kernel/pull/2813/files#diff-5038dc48317648056e713d64d82734beddb5656c0bfe16bd60629f183430d68aL95
should be BQK * N (to account for the whole matrix)
6. ```struct GemmConfigRowColQuant : public GemmConfigBase``` typo in name

amd-khushbu committed:
addressing review comments

CongMa13 committed:
[CK TILE GEMM] Replace get_preshuffle_or with is_quantpreshuffle_enabled

The get_preshuffle_or was not working as expected, which led to incorrect behavior
in the quantization preshuffle process. This change replaces it with the more reliable
is_quantpreshuffle_enabled function to properly determine when preshuffle should be applied.

amd-khushbu committed:
Resolved merge conflicts

amd-khushbu committed:
Resolved merge conflicts
